### PR TITLE
util,requestbatcher: prevent goroutine leaks in test functions

### PR DESCRIPTION
**Pull request by lzhfromustc**
***Description***
There are 4 test functions sharing a similar problem: some goroutines in these functions will be leaked in certain cases (like timeout). 

Although `t.Fatal()` is often called in these certain cases, it won't stop the leaking, because ["Calling FailNow does not stop those other goroutines." ](https://golang.org/pkg/testing/#T.FailNow)

***How they are fixed***
The root cause of all these problems is the blocking of channel operations. This patch makes sure that all these operations will be unblocked if test fails.

Other fixes are straightforward, but `TestBackpressure()` is a little bit more complex because `reply()` is called 3 times, and only the third time may be leaked. I used a timer to unblock it after enough time.

### CODE REVIEW

**Review by ajwerner — CHANGES_REQUESTED** [review]
Out of curiosity, how did you come to encounter these if they can only happen if the tests fail? 

I’m not yet convinced that they constitute a problem. Having a test hang when it’s expectations are violated isn’t the worst failure mode. 

> Reviewed 1 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42094#-:-Ls_6X44-CZ5VdN2Tg8g:bsa81tj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lzhfromustc)

___
*[pkg/internal/client/requestbatcher/batcher_test.go, line 186 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42094#-Ls_51L4BOT2dpSrV2ni:-Ls_51L57bh6AddBD-VK:b6ku79o) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3705c0a79d0f0f9d7bedc3ebec6adab69e8a403/pkg/internal/client/requestbatcher/batcher_test.go#L186)):*
> ```Go
> 	runtime.Gosched() // tickle the runtime in case there might be a timing bug
> 	assert.Equal(t, int64(0), atomic.LoadInt64(&sent))
> 	canReply <- struct{}{} // now the two requests should send
> ```

One question I have is whether this leak is possible if the test passes? If the answer is no, why do we care about the leak? Did you encounter it by introducing a bug to the code and then observed it failing? Or did you run with a very short timeout and then found it was hung?
___
*[pkg/internal/client/requestbatcher/batcher_test.go, line 188 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42094#-Ls_5R3F0z61GLkRErAp:-Ls_5R3F0z61GLkRErAq:bi5pge0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3705c0a79d0f0f9d7bedc3ebec6adab69e8a403/pkg/internal/client/requestbatcher/batcher_test.go#L188)):*
> ```Go
> 	canReply <- struct{}{} // now the two requests should send
> 	go func(){
> 		time.Sleep(50 * time.Second) // SucceedsSoon's duration is at most 45 seconds
> ```

This is pretty gross. If we’re going to do something here, how about:

```
defer func() {
    select {
    case <-canReply:
    default:
        close(canReply)
    }
}()
```

**Review by lzhfromustc — COMMENTED** [review]
Thanks for your review! I am developing a set of static checkers to find concurrency bugs. One checker is to find goroutine leaks, and it can only find these 4 test function in cockroach. 
You are right. Goroutine leaks in test are not big problems, but still these goroutines will consume some memory. If we can easily make them "unblockable", then I suppose we should do it.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42094#-:-Ls_8Ue0DveYK2n0soSy:bp961q1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/internal/client/requestbatcher/batcher_test.go, line 186 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42094#-Ls_51L4BOT2dpSrV2ni:-Ls_AJddDNwCm_2GDx9Z:b2azexu) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3705c0a79d0f0f9d7bedc3ebec6adab69e8a403/pkg/internal/client/requestbatcher/batcher_test.go#L186)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

One question I have is whether this leak is possible if the test passes? If the answer is no, why do we care about the leak? Did you encounter it by introducing a bug to the code and then observed it failing? Or did you run with a very short timeout and then found it was hung?
</blockquote></details>

No, these leaks won't happen if the test pass. However, if tests don't pass, leaked goroutines will still consume memory, because `testutils.SucceedsSoon()` calls `t.Fatal()` and then `runtime.Goexit()`, which will only kill the current goroutine.
___
*[pkg/internal/client/requestbatcher/batcher_test.go, line 188 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42094#-Ls_5R3F0z61GLkRErAp:-Ls_B3Z96b3SHJymWRPK:bhsfgfb) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3705c0a79d0f0f9d7bedc3ebec6adab69e8a403/pkg/internal/client/requestbatcher/batcher_test.go#L188)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This is pretty gross. If we’re going to do something here, how about:

```
defer func() {
    select {
    case <-canReply:
    default:
        close(canReply)
    }
}()
```
</blockquote></details>

This is truly gross! I hesitated for how I should fix this bug. I don't think my patch is a good one, because the timer is also consuming the memory, but `defer` doesn't work here, because the `t.Fatal()` in `testutils.SucceedsSoon()` won't run `defer`. I always think the design of `t.Fatal()` is bug-prone. Anyway, maybe we can ignore this leak here, since fixing it may even consume more memory.

**Review by ajwerner — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42094)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lzhfromustc)

___
*[pkg/internal/client/requestbatcher/batcher_test.go, line 186 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42094#-Ls_51L4BOT2dpSrV2ni:-Ls_MYRt348E_FTkdisw:beq3xuz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3705c0a79d0f0f9d7bedc3ebec6adab69e8a403/pkg/internal/client/requestbatcher/batcher_test.go#L186)):*
<details><summary><i>Previously, lzhfromustc wrote…</i></summary><blockquote>

No, these leaks won't happen if the test pass. However, if tests don't pass, leaked goroutines will still consume memory, because `testutils.SucceedsSoon()` calls `t.Fatal()` and then `runtime.Goexit()`, which will only kill the current goroutine.
</blockquote></details>

As soon as the test fails the process is going to exit so I don’t much care about the memory usage. Furthermore go look at that `defer leaktest.AfterTest(t)()` at the top of every test in this repo. It will fail the test if the set of goroutines at the end is different from at the start. We have a linter forcing the inclusion of this call. 

I appreciate your using your tool on our repo. I need to think more before I decide whether I’m interested in doing anything in these test cases it has uncovered. 
___
*[pkg/internal/client/requestbatcher/batcher_test.go, line 188 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42094#-Ls_5R3F0z61GLkRErAp:-Ls_McB8DHSlTDcpGvzh:b-ovdeiw) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3705c0a79d0f0f9d7bedc3ebec6adab69e8a403/pkg/internal/client/requestbatcher/batcher_test.go#L188)):*
<details><summary><i>Previously, lzhfromustc wrote…</i></summary><blockquote>

This is truly gross! I hesitated for how I should fix this bug. I don't think my patch is a good one, because the timer is also consuming the memory, but `defer` doesn't work here, because the `t.Fatal()` in `testutils.SucceedsSoon()` won't run `defer`. I always think the design of `t.Fatal()` is bug-prone. Anyway, maybe we can ignore this leak here, since fixing it may even consume more memory.
</blockquote></details>

Deferred calls are run with Goexit. SucceedsSoon runs on the caller’s goroutine. 

Goexit terminates the goroutine that calls it. No other goroutine is affected. Goexit runs all deferred calls before terminating the goroutine. 

https://golang.org/pkg/runtime/#Goexit

**Review by lzhfromustc — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42094)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/internal/client/requestbatcher/batcher_test.go, line 186 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42094#-Ls_51L4BOT2dpSrV2ni:-LscDwXc4HB7D2gci12H:b-qo7s4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3705c0a79d0f0f9d7bedc3ebec6adab69e8a403/pkg/internal/client/requestbatcher/batcher_test.go#L186)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

As soon as the test fails the process is going to exit so I don’t much care about the memory usage. Furthermore go look at that `defer leaktest.AfterTest(t)()` at the top of every test in this repo. It will fail the test if the set of goroutines at the end is different from at the start. We have a linter forcing the inclusion of this call. 

I appreciate your using your tool on our repo. I need to think more before I decide whether I’m interested in doing anything in these test cases it has uncovered. 
</blockquote></details>

Fascinating, I have never noticed the usage of `defer leaktest.AfterTest(t)()`. I just took a look at similar goroutine leaks in test in Docker and Kubernetes, and they haven't applied anything like this yet, but etcd does in some test functions. 
With this, these goroutine leaks won't be a problem. Whether to handle these leaks or not may only effect whether there will be more error messages during `AfterTest`.
___
*[pkg/internal/client/requestbatcher/batcher_test.go, line 188 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42094#-Ls_5R3F0z61GLkRErAp:-LscD3309rvd9EgMTc75:btu52do) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3705c0a79d0f0f9d7bedc3ebec6adab69e8a403/pkg/internal/client/requestbatcher/batcher_test.go#L188)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Deferred calls are run with Goexit. SucceedsSoon runs on the caller’s goroutine. 

Goexit terminates the goroutine that calls it. No other goroutine is affected. Goexit runs all deferred calls before terminating the goroutine. 

https://golang.org/pkg/runtime/#Goexit
</blockquote></details>

Thanks you very much for pointing this out! I could have written better patch with this in mind.

**Comment by lzhfromustc** [comment]
Sorry, I haven't been thinking of this PR for quite some time...
I just edited my commit. Now I use a defer to handle the case `testutils.SucceedsSoon()` calls `t.Fatal()`. If `t.Fatal()` is called, then `if t.Failed() {}` will close `canReply`, so the third goroutine running `reply()` will return.
Still, whether we fix this or not is your call.

**Review by ajwerner — APPROVED** [review]
:lgtm: mod the linter failure. Once the build is fixed I'll merge it. 

Thanks for coming back to this!

> Reviewed 1 of 3 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42094#-:-M0-2Ztw7baEShlWh6TU:b-mcpsdb)* status: :shipit: complete! 1 of 0 LGTMs obtained

**Comment by lzhfromustc** [comment]
I think now the linter is fixed. I am not sure what is "bors" checker. Can't see its result

**Review by ajwerner — COMMENTED** [review]
Thanks for your contribution! Bors is the bot we use to merge commits https://bors.tech/

bors r=ajwerner

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42094#-:-M1CRD8-DywrjquI-9Xv:b6zy4k0)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner)

**Comment by lzhfromustc** [comment]
I tried to see the results of bors, but seems I don't have the permission...

**Review by ajwerner — COMMENTED** [review]
The build flaked on something unrelated.

bors r=ajwerner

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42094#-:-M1_Xg3O5ZjcHPXAbHW6:b-eidbh7)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner)

**Pull request merged**
